src/java/org/apache/cassandra/service/reads/ReadCallback.java

```diff
@@ -99,9 +99,12 @@ public boolean await(long timePastStart, TimeUnit unit)
     public void awaitResults() throws ReadFailureException, ReadTimeoutException
     {
         boolean signaled = await(command.getTimeout(MILLISECONDS), TimeUnit.MILLISECONDS);
-        boolean failed = failures > 0 && blockFor + failures > replicaPlan().contacts().size();
+        boolean failed = failures > 0 && (blockFor > resolver.responses.size() || !resolver.isDataPresent());
         if (signaled && !failed)
+        {
+            assert resolver.isDataPresent();
```

Comment:
can you add a assert message?

Reply:
Maybe we should just take the assertion already in `getData()` and move it to `AbstractReadExecutor#awaitResponses()`, immediately after the call to `awaitResults()`? It would still cover all the code paths and not require us to make it twice for some.